h2o-py/h2o/plot/_plot_result.py

```diff
@@ -0,0 +1,26 @@
+# -*- encoding: utf-8 -*-
+# mutable versions of py immutable types
+class _MObject(object): pass
+class _MTuple(tuple): pass
+class _MList(list): pass
+class _MDict(dict): pass
+class _MStr(str): pass
+
+def decorate_plot_result(res=None, figure=None):
+    # list all special/immutable types that we need first
+    if res is None:
+        dec = _MObject()
+    elif isinstance(res, tuple):
+        dec = _MTuple(res)
+    elif isinstance(res, list):
+        dec = _MList(res)
+    elif isinstance(res, dict):
+        dec = _MDict(res)
+    elif isinstance(res, str):
+        dec = _MStr(res)
+    else: # should be an H2O instance, should be mutable
+        dec = res
+    dec.figure = figure
+    return dec
+
```

Comment:
can you please add a test for this, covering all basic use-cases? just to be sure it works exactly as expected, thx

```python
figure = "let's pretend I'm a figure"

res = decorate_plot_result(figure=figure)
assert res.figure == figure

res = decorate_plot_result((3, 5), figure=figure)
assert isinstance(res, tuple)
a, b = res
assert a == 3
assert b == 5
assert res.figure == figure

res = decorate_plot_result([3, 5, 6], figure=figure)
assert isinstance(res, list)
assert res == [3, 5, 6]
assert res.figure == figure

# same with dict

 class Foo(object):
    def __init__(self, bar=None):
        self.bar=bar

res = decorate_plot_result(Foo(bar="baz"), figure=figure)
assert isinstance(res, Foo)
assert res.bar == "baz"
assert res.figure == figure
```

Reply:
tests are added on places where underlying plot functions are being tested. But can add also test like this

Reply:
d535954

Reply:
thanks, I just don't understand why it's tested in a GLM test file...
`h2o/module/` -> `[ tests/testdir_module/pyunit_test_a.py, tests/testdir_module/pyunit_test_b.py, …]`
or at least  `tests/testdir_misc/pyunit_test_module.py`

I know that the legacy file structure is messy and annoying, but it's not a reason to keep using it.

Imagine someone adding a function in that module, do you think this person will search for existing tests in glm?

Reply:
will move all of them to pyunit_PUBDEV-8385_add_saving_parameters_to_plot_functions.py

Reply:
are you putting `PUBDEV-xxxx` in all your tests file names?
I don't know who came with this first.
When a ticket is closed, it should be… closed. And we shouldn't care why a test was written: a test is always written to test existing code, no matter if a ticket was previously filed for it (because it wasn't tested…) or just because it's code, so whenever possible, it should be tested.

Reply:
not to all.. I think it's good to know the ticket number because you can easily find underlying jira -> underlying support ticket -> underlying information why something was done. It can be found in a commit message but this can get lost after some refactoring, so I don't think it's such a bad thing to have a ticket number in a test name.

Reply:
but again, you have a `_plot_result.py` module. In one month, you modify or extend this module, will you rather search for `pyunit_plot_result` or `pyunit_pubdev-8385`??
The pb with `pubdev-xxx` in the name is that it makes sense only for the person who writes the test, and only at the time of the PR, for others and later it's just meaningless noise.